lib/spack/spack/spec.py

```diff
@@ -2542,9 +2548,14 @@ def _mark_concrete(self, value=True):
         Only for internal use -- client code should use "concretize"
         unless there is a need to force a spec to be concrete.
         """
+        # if set to false, clear out all hashes (set to None or remove attr)
+        # may need to change references to respect None
         for s in self.traverse():
             if (not value) and s.concrete and s.package.installed:
                 continue
+            elif (not value):
+                s._hash = None
+                s._full_hash = None
```

Comment:
there's also the build hash here.

Reply:
So it turns out that this condition is most likely flawed.
(I tested by adding a call to the new clear_hashes function in here and it broke pretty much everything.
Since this isn't needed for splice, I am removing it for now.

Reply:
The cache clearing logic issue was resolved in an offline conversation.